Search crashes when searching last letter of filename containing İ

### Summary

<!-- Begin your issue with a one sentence summary -->
When a codebase contains a file with the letter İ in its name (U+0130 Latin Capital Letter I with Dot Above), typing the last letter of that filename into the Ctrl+P file search box crashes the editor.

### Description
<!-- Include all steps necessary to reproduce from a clean Zed installation. Be verbose -->
Steps to trigger the problem:
1. Create a file named `İg` in a project
2. Open the file switcher (Ctrl+P)
3. Type `g`

Actual Behavior: Zed hangs for a second or two and closes
Expected Behavior: It searches for files with `g` in their name


### Zed Version and System Specs

Zed: v0.183.11 (Zed) 
OS: Linux Wayland opensuse-tumbleweed 20250426
Memory: 15.2 GiB
Architecture: x86_64
GPU: Intel(R) Iris(R) Xe Graphics (RPL-P) || Intel open-source Mesa driver || Mesa 25.0.4

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

<details><summary>Zed.log</summary>

The log that seems to be relevant:

```log
2025-04-20T23:04:30.134605705+03:00 [ERROR] {
  "thread": "<unnamed>",
  "payload": "index out of bounds: the len is 2 but the index is 2",
  "location_data": {
    "file": "crates/fuzzy/src/matcher.rs",
    "line": 224
  },
  "backtrace": [
    "zed::reliability::init_panic_hook::{{closure}}::hba4876f7a43de794+141243845",
    "std::panicking::rust_panic_with_hook::h541791bcc774ef34+115626131",
    "std::panicking::begin_panic_handler::{{closure}}::h6479a2f0137c7d19+115625354",
    "std::sys::backtrace::__rust_end_short_backtrace::ha04e7c0fc61ded91+115618873",
    "rust_begin_unwind+115624493",
    "core::panicking::panic_fmt::h5764ee7030b7a73d+8655904",
    "core::panicking::panic_bounds_check::h0328ca7e7f0749c4+8656386",
    "fuzzy::matcher::Matcher::recursive_score_match::hc91af778db1db651+60440043",
    "fuzzy::matcher::Matcher::score_match::h6f5a71762b197898+60437365",
    "gpui::executor::Scope::spawn::{{closure}}::h6c0f35aea243abc8+48450412",
    "async_task::raw::RawTask<F,T,S,M>::run::h70f9161f6c4dbede+45970218",
    "std::sys::backtrace::__rust_begin_short_backtrace::h470febbd6b286bf8+65227187",
    "core::ops::function::FnOnce::call_once{{vtable.shim}}::h27790ad43661da79+65346996",
    "std::sys::pal::unix::thread::Thread::new::thread_start::hcc5ed016d554f327+115656219",
    "start_thread+45445581780294",
    "__GI___clone3+45445582321996"
  ],
  "app_version": "0.182.11",
  "app_commit_sha": "67ca67273f4066b703e3bed9d0a1a08b13010c43",
  "release_channel": "stable",
  "target": "x86_64-unknown-linux-gnu",
  "os_name": "Linux Wayland",
  "os_version": "opensuse-tumbleweed 20250417",
  "architecture": "x86_64",
  "panicked_on": 1745179470134,
  "system_id": "eafc1df2-bd9a-4c16-ae17-fe004959edc4",
  "installation_id": "45ccdf80-15ab-4133-8c49-17d59ae5c0e3",
  "session_id": "7fb53a8e-c548-46d0-bc97-b2099520e678"
}
```

</details>

--- a/crates/file_finder/src/file_finder_tests.rs
+++ b/crates/file_finder/src/file_finder_tests.rs
@@ -243,6 +243,38 @@ async fn test_matching_paths(cx: &mut TestAppContext) {
 }
 
 #[gpui::test]
+async fn test_unicode_paths(cx: &mut TestAppContext) {
+    let app_state = init_test(cx);
+    app_state
+        .fs
+        .as_fake()
+        .insert_tree(
+            path!("/root"),
+            json!({
+                "a": {
+                    "İg": " ",
+                }
+            }),
+        )
+        .await;
+
+    let project = Project::test(app_state.fs.clone(), [path!("/root").as_ref()], cx).await;
+
+    let (picker, workspace, cx) = build_find_picker(project, cx);
+
+    cx.simulate_input("g");
+    picker.update(cx, |picker, _| {
+        assert_eq!(picker.delegate.matches.len(), 1);
+    });
+    cx.dispatch_action(SelectNext);
+    cx.dispatch_action(Confirm);
+    cx.read(|cx| {
+        let active_editor = workspace.read(cx).active_item_as::<Editor>(cx).unwrap();
+        assert_eq!(active_editor.read(cx).title(cx), "İg");
+    });
+}
+
+#[gpui::test]
 async fn test_absolute_paths(cx: &mut TestAppContext) {
     let app_state = init_test(cx);
     app_state

--- a/crates/fuzzy/src/matcher.rs
+++ b/crates/fuzzy/src/matcher.rs
@@ -1,5 +1,6 @@
 use std::{
     borrow::{Borrow, Cow},
+    collections::BTreeMap,
     sync::atomic::{self, AtomicBool},
 };
 
@@ -50,7 +51,7 @@ impl<'a> Matcher<'a> {
 
     /// Filter and score fuzzy match candidates. Results are returned unsorted, in the same order as
     /// the input candidates.
-    pub fn match_candidates<C, R, F, T>(
+    pub(crate) fn match_candidates<C, R, F, T>(
         &mut self,
         prefix: &[char],
         lowercase_prefix: &[char],
@@ -65,6 +66,7 @@ impl<'a> Matcher<'a> {
     {
         let mut candidate_chars = Vec::new();
         let mut lowercase_candidate_chars = Vec::new();
+        let mut extra_lowercase_chars = BTreeMap::new();
 
         for candidate in candidates {
             if !candidate.borrow().has_chars(self.query_char_bag) {
@@ -77,9 +79,14 @@ impl<'a> Matcher<'a> {
 
             candidate_chars.clear();
             lowercase_candidate_chars.clear();
-            for c in candidate.borrow().to_string().chars() {
+            extra_lowercase_chars.clear();
+            for (i, c) in candidate.borrow().to_string().chars().enumerate() {
                 candidate_chars.push(c);
-                lowercase_candidate_chars.append(&mut c.to_lowercase().collect::<Vec<_>>());
+                let mut char_lowercased = c.to_lowercase().collect::<Vec<_>>();
+                if char_lowercased.len() > 1 {
+                    extra_lowercase_chars.insert(i, char_lowercased.len() - 1);
+                }
+                lowercase_candidate_chars.append(&mut char_lowercased);
             }
 
             if !self.find_last_positions(lowercase_prefix, &lowercase_candidate_chars) {
@@ -97,6 +104,7 @@ impl<'a> Matcher<'a> {
                 &lowercase_candidate_chars,
                 prefix,
                 lowercase_prefix,
+                &extra_lowercase_chars,
             );
 
             if score > 0.0 {
@@ -131,18 +139,20 @@ impl<'a> Matcher<'a> {
     fn score_match(
         &mut self,
         path: &[char],
-        path_cased: &[char],
+        path_lowercased: &[char],
         prefix: &[char],
         lowercase_prefix: &[char],
+        extra_lowercase_chars: &BTreeMap<usize, usize>,
     ) -> f64 {
         let score = self.recursive_score_match(
             path,
-            path_cased,
+            path_lowercased,
             prefix,
             lowercase_prefix,
             0,
             0,
             self.query.len() as f64,
+            extra_lowercase_chars,
         ) * self.query.len() as f64;
 
         if score <= 0.0 {
@@ -173,12 +183,13 @@ impl<'a> Matcher<'a> {
     fn recursive_score_match(
         &mut self,
         path: &[char],
-        path_cased: &[char],
+        path_lowercased: &[char],
         prefix: &[char],
         lowercase_prefix: &[char],
         query_idx: usize,
         path_idx: usize,
         cur_score: f64,
+        extra_lowercase_chars: &BTreeMap<usize, usize>,
     ) -> f64 {
         use std::path::MAIN_SEPARATOR;
 
@@ -200,15 +211,22 @@ impl<'a> Matcher<'a> {
 
         let mut last_slash = 0;
         for j in path_idx..=limit {
-            let path_char = if j < prefix.len() {
+            let extra_lowercase_chars_count = extra_lowercase_chars
+                .iter()
+                .take_while(|(i, _)| i < &&j)
+                .map(|(_, increment)| increment)
+                .sum::<usize>();
+            let j_regular = j - extra_lowercase_chars_count;
+
+            let path_char = if j_regular < prefix.len() {
                 lowercase_prefix[j]
             } else {
-                path_cased[j - prefix.len()]
+                path_lowercased[j - prefix.len()]
             };
             let is_path_sep = path_char == MAIN_SEPARATOR;
 
             if query_idx == 0 && is_path_sep {
-                last_slash = j;
+                last_slash = j_regular;
             }
 
             #[cfg(not(target_os = "windows"))]
@@ -218,18 +236,18 @@ impl<'a> Matcher<'a> {
             #[cfg(target_os = "windows")]
             let need_to_score = query_char == path_char || (is_path_sep && query_char == '_');
             if need_to_score {
-                let curr = if j < prefix.len() {
-                    prefix[j]
+                let curr = if j_regular < prefix.len() {
+                    prefix[j_regular]
                 } else {
-                    path[j - prefix.len()]
+                    path[j_regular - prefix.len()]
                 };
 
                 let mut char_score = 1.0;
                 if j > path_idx {
-                    let last = if j - 1 < prefix.len() {
-                        prefix[j - 1]
+                    let last = if j_regular - 1 < prefix.len() {
+                        prefix[j_regular - 1]
                     } else {
-                        path[j - 1 - prefix.len()]
+                        path[j_regular - 1 - prefix.len()]
                     };
 
                     if last == MAIN_SEPARATOR {
@@ -279,17 +297,18 @@ impl<'a> Matcher<'a> {
 
                 let new_score = self.recursive_score_match(
                     path,
-                    path_cased,
+                    path_lowercased,
                     prefix,
                     lowercase_prefix,
                     query_idx + 1,
                     j + 1,
                     next_score,
+                    extra_lowercase_chars,
                 ) * multiplier;
 
                 if new_score > score {
                     score = new_score;
-                    best_position = j;
+                    best_position = j_regular;
                     // Optimization: can't score better than 1.
                     if new_score == 1.0 {
                         break;

